Fix system boot time calculation

eopo opened the pull request:
Fix #378 - When calculating the systems boot time, the uptime was subtracted from the current timestamp. The timestamp included milliseconds, the uptime not. This resulted in the calculated boot time's milliseconds being purely dependent on the exact timing of the request. The timestamp seems to be rounded afterwards, resulting in it switching between two second values depending on the request taking place in the first half or the second half of a second.

This PR set's the current timestamps milliseconds to zero, resulting in the calculated boot times ms always being zero as well. Rounding should therefore always result in the same value.

alexdelprete commented:
We could also round off the microseconds in the result:

```python
system["boottime"] = boottime.replace(microsecond=0).timestamp()
```

I was also thinking that boot time is derived indirectly from uptime, because API is missing the specific value. I wonder if it would be better to use `psutil.boot_time()` so we don't have to do all those calculations.

Snuffy2 commented:
I’m out of town on vacation this week but I will review next week then I return. Thanks for the submission!

eopo commented:
> We could also round off the microseconds in the result
I think it doesn't matter. The timedelta doesn't include µs anyways, so if they get removed from the timestamp or the result should be the same.



> I wonder if it would be better to use something like `psutil.boot_time()` at OS level so we don't have to do those calculations/conversions

That's out of the scope I know how to navigate. If you think that would be a better way, let's go. The way I wrote it seemed the easiest for me.

alexdelprete commented:
> That's out of the scope I know how to navigate. If you think that would be a better way, let's go.

It was more an idea for @Snuffy2. :)

> The way I wrote it seemed the easiest for me.

It's very subjective: in general, I always do the rounding after the calculations, on the result, just in case some intermediate steps reintroduce something. In this case there's no difference.

eopo commented:
Test system runs fine, but i've had another thought about it and it might be, that this way won't reliably fix the issue.
OPNsense itself would be applying some rounding, and requesting the value a few µs apart (within the same second from HA's point of view) could deliver different uptimes. 

It might be, that the PHP 'Injection' is the only independent reliable way at this time.

I've implemented exposing the boot time in the OPNsense API and opened a Pull Request https://github.com/opnsense/core/issues/8556 for that - let's see, if they accept it. If so, we could consume this field instead. For backwards compatibility, I'd propose to keep the current way too and differentiate by OPNsense Version, which way to use.

alexdelprete commented:
Like I said, the proper way is having an API boottime, or we can get it through the OS. I hope devs accept your PR. Great job.

I noticed in the API response this `config` value:

```json
{
    "uptime": "17:18:44",
    "datetime": "Tue Apr 15 18:12:22 CEST 2025",
    "config": "Tue Apr 15 1:17:31 CEST 2025",
    "loadavg": "1.15, 0.59, 0.41"
}
```

At a first glance it seemed almost like the boot time (datetime-uptime), but it's not exactly correct. From your PR it seems to be last config. update.

eopo commented:
It happened again on the regular version, my patched version didn't show the same error - we can't be too sure though, because they might be refreshed on different times.

eopo commented:
I thought about it again and realized that it makes no difference.

The OPNsense API does not round, but ignores the µs, which effectively equals a floor.
Our real problem is the missing information about the µs together with the fact that the relevant times (system time OPNsense, system time Home Assistant and uptime string OPNsense) are not synchronized regarding the second changes.
Even if we ignore the µs on our side, this means that within one second of system time (regardless of whether OPNsense or HA) two different values can be returned for the uptime, depending on whether the second change of the uptime has already taken place or not.

The result is: For a reliable calculation of the boot time, we either need the µs of the uptime or simply the timestamp of the boot time itself - with the latter, of course, being a much more elegant way. We must therefore either hope that OPNsense accepts the PR or find a solution ourselves. In addition to the possibility of remote PHP execution, I think it would also be possible to write a plugin that generally delivers all the values required for HA that are missing from the official API as an independent API route. 

Translated with DeepL.com (free version)

alexdelprete commented:
> Our real problem is the missing information about the µs together with the fact that the relevant times (system time OPNsense, system time Home Assistant and uptime string OPNsense) are not synchronized regarding the second changes.

I just realized that the uptime is relative to the OPNsense system time, not HA's system time. Clocks might not be synced so we're doing the wrong calculations.

So instead of doing:

boottime = (HA system time) - (OPNsense uptime)

we should do:

OPNsense_boottime = OPNsense_datetime - OPNsense_uptime

`datetime` and `uptime` are in the API endpoint response, and have a precision up to seconds, so no microseconds round-off is needed. The calculation should always provide the same value.

> In addition to the possibility of remote PHP execution, I think it would also be possible to write a plugin that generally delivers all the values required for HA that are missing from the official API as an independent API route.

That's what the plugin you install in OPNsense for the integration does: it allows the integration to fetch values through PHP that are not available in the API.

alexdelprete commented:
> This shows, that the calculated boot time is alternating between 0 (0.5) and 1, depending on the timing of the request - Even if we use the OPNsense system time as reference, and even if we remove µs from the calculation at all.

I don't understand that table at all, the difference between the first 2 columns is **always** 0.5, it doesn't depend on when you do the calculation because time is constant for both variables and the time source is the same:

![image](https://github.com/user-attachments/assets/6738a0c1-1367-48dc-ba43-bafb459ad142)

I really don't see how the difference between the 2 values changes from 0 to 1 like you show in the table, the time is constant for both system time and uptime, so if it starts with 0.5 of difference it will be 0.5 for all the rest of the samples. Unless we assume the OPNsense system will be traveling at speed of light and the observer is not. 🤣 

If system time and boot time are stored with µs precision, let's eliminate the µs and convert them to unix timestamps, and only after that we calculate the difference. Then we convert the result in datetime.

> Good news. The PR at OPNsense was accepted, so the API should be outputting the Systems Boot Time in the format it is currently outputting the System Time as well. 🥳 I don't know how fast it will be available in releases though.

That is AWESOME. :)

alexdelprete commented on custom_components/opnsense/pyopnsense/__init__.py:
```diff
@@ -1272,20 +1272,26 @@ async def _get_telemetry_system(self) -> MutableMapping[str, Any]:
+                seconds = OPNsenseClient._try_to_int(seconds_str, 0) or 0
+                system["uptime"] = days * 86400 + hours * 3600 + minutes * 60 + seconds
+
+                boottime: datetime = datetime.now() - timedelta(seconds=system["uptime"])
```
Like I said in the discussion, this is not correct: we shouldn't use HA system time but `datetime` from the API endpoint response.

eopo commented:
I thought I found an error in your assumption, but I guess I let one fact out.
OPNsense disregards the ms on the time formatting for systemtime and disregards the ms on boot time -> The seconds are always in sync. So yeah, if we use the OPNsense system time, it should be fine. I'll add it.

eopo commented:
The only thing i'm not 100% sure on that: If for whatever reason I decided to set my OPNsenses system time to the 1.1.2000 00:00 and boot it that day, HA would show the boot time the same way - Whilst the HA does operate with correct timing.
If we calculate the timedelta between OPNsense and HA and add/subtract it to the boottime, we can ensure that the boottime representation in HA is in the same time reference as the HA system time.

alexdelprete commented:
> If we calculate the timedelta between OPNsense and HA and add/subtract it to the boottime, we can ensure that the boottime representation in HA is in the same time reference as the HA system time.

I see what you mean, but I don't agree: the entity has to represent what the value is in the observed system, it doesn't have to be adapted/modified.

Let me make a provocation for the sake of clearing my thought for you: when OPNsense will provide `boottime` in the API and we get it from there, and let's suppose the clock in OPNsense is not in sync with HA's clock, what do you suggest to do? Fix the `boottime`? I don't think so...you need to get data from the system and present it in the HA entity, without manipulation.

You need to fix the clock in OPNsense (use chrony or ntp), that is the root of the problem, not the data.

Hope I've managed to be clear enough. :)

alexdelprete commented on custom_components/opnsense/pyopnsense/__init__.py:
```diff
@@ -1273,10 +1273,12 @@ async def _get_telemetry_system(self) -> MutableMapping[str, Any]:
             boottime: datetime = datetime.strptime(time_info["boottime"], "%a %b %d %H:%M:%S %Z %Y")
             system["boottime"] = boottime.timestamp()
-            system["uptime"] = int(datetime.now() - system["boottime"])
+            system["uptime"] = int(systemtime - system["boottime"])
```
uptime is available in the API response, it shouldn't be calculated.

eopo commented:
@alexdelprete  My brain tricked me into assuming, that the uptime was calculated precisely in OPNsense, which would have led to the effects I mentioned being the case,

But you are right. The OPNsense subtracts kind of constant whole seconds from a clock floored to the last second, so it only depends on the system time second switch.

Regarding the second thought I had: I think it's a personal preference thing on how you look at the entity. If you say "This is what OPNsense thinks it's boot time is", you are right (Multiple OPNsense with different system time started at the same moment would show different boot times in HA). If you say "This is when the system was booted, in the same reference time as all other times in Home Assistant" I am right (Multiple OPNsense with different system time started at the same moment would show the same boot time in HA). I'm fine with either.

---

@travisghansen Thats what we're doing now, The integration had this uptime field before, so I implemented it for the new way as well. I don't think it is exposed as sensor, but since I don't know if it is used somewhere else, I didn't omit it.

eopo commented:
Oh, and OPNsense told us that the patch would be included in an upcoming minor release, so I think it's gonna be a few weeks up to a few months.

alexdelprete commented:
> I’m not sure what all is happening in the PR but just a reminder that using uptime as a sensor is less desirable than boot time from the perspective of storage in the db and having a sensor that isn’t constantly updating. Carry on :)

My friend, how are you? Long time no see. Hope everything is fine. 

We're just discussing on time issues, and as often happens, brains get excited and start traveling in different dimensions. But I think we finally got to a common conclusion.  :)

alexdelprete commented:
> I think it's a personal preference thing on how you look at the entity.

We're almost done, we just need to clarify this point. 

The way I look at it is that an entity in HA that represents data of another system, shouldn't be altered in any way. Transformations should be up to the user, the integration should provide exact data. 

We could also provide unaltered boottime as main sensor value and adjusted boottime as an attribute of the sensor, so users can pick which one they want. 

Let's wait for @Snuffy2 to chime in. :)

eopo commented:
> We could also provide unaltered boottime as main sensor value and adjusted boottime as an attribute of the sensor, so users can pick which one they want.

Uhhh, i like that thought. 

Anyhow, i just implemented a two step logic.
- If the boottime is available in API, use it for boottime - Else calculate from uptime and systemtime
- If the uptime is parsed from API, use it for uptime - Else calculate from boottime and systemtime (In case for whatever reason the parsing doesn't work)
- If neither is working, act like before (Warn log)

I had a look, it seems like the uptime property is not used somewhere else. It was explicitly set as `system["uptime"]`, so I chose to keep it. If we are shure we don't need it, i think we could easily remove it.

I think that gives us the best of both worlds.

eopo commented:
It is stored into that key, but never exposed as Sensor - so currently, there seems no way for the user to access it.

alexdelprete commented:
> It is stored into that key, but never exposed as Sensor - so currently, there seems no way for the user to access it.

I would create a sensor for uptime, disabled by default. Uptime is more useful, but like Travis said it is updated constantly. But if we leave it disabled it will be up to the user to decide. I would enable it, but it's personal preference.

Let's wait for @Snuffy2 to chime in.

eopo commented:
Yeah, sounds good to me! I'll build it as a separate PR

eopo commented:
Happy news: My pr to OPNsense was now included in the release. I think we can merge this one for here as well, to make use of that.

alexdelprete commented:
> Unfortunately, implementing a default-deactivated sensor would need a seperate Sensor class. It's just not worth the work for me for a sensor,

We already have entities that are disabled by default in the code.

Snuffy2 commented:
I'm very sorry for the delay. 

Looks good and great job getting boottime added to opnsense as well. I made a couple of small refinements. Will merge and release shortly.
